New option and bug fix

ScoobyDid opened the pull request:
<!--
Thanks for creating this pull request!

Please make sure that the pull request is limited to one type (docs, feature, etc.) and keep it as small as possible. You can open multiple PRs instead of opening a huge one.
-->

**Description**

- Fixed bug when screen reader sometimes did not pronounce announcement due to 0 timeout (tested on NVDA on windows 11)
- Added new option - delay, that defaults to 100
- Updated `announce` method to include delay parameter

**Checks**

<!--
Make sure the PR fulfills as many of the following requirements as possible
-->

- [x] The PR is submitted to the `master` branch
- [x] The code was linted before pushing (`npm run lint`)
- [x] All tests are passing (`npm run test`)
- [ ] New or updated tests are included
- [x] The documentation was updated as required

hirasso commented:
Amazing work, thank you @ScoobyDid for debugging and creating this PR! Based on your reported test results, the current implementation could almost be called buggy.

I'm unsure if this should be adjustable as an option – what is your reasoning behind this? Setting `delay` to `0` would just re-introduce the flaky behavior between screen readers. Also, a delay of 100ms wouldn't be noticeable to end users, so I'd prefer to hard-code this. @daun what do you think?

daun commented:
Nice. Didn't know the live region package had a delay option. Adding a default delay here makes sense. Was 100ms the lowest value to reliably fix the issue? I'd try to go as low as possible without reintroducing the issue.

ScoobyDid commented:
@daun the problem is that there is no specific threshold that is 100% enough for screen reader to pronounce the updated text - it depends on the screen reader - how busy it is at the moment, 
and on user device. this is especially true when user navigates to a different page - there may be too many changes for screen reader to process. On my machine 50ms and 30ms works,
but it may not work for all users. 100ms should be enough to almost 100% guarantee that the text will be pronounced.

ScoobyDid commented:
@hirasso Nice point, i didn't really think about it. I thought it would be a good idea to allow users to fine-tune the delay based on their specific needs and the behavior of their application.
However, you're correct that this could potentially introduce unwanted behavior if a user sets the delay to 0ms or other low value.

To solve this, i can update this PR to do one of following:

1. As you suggested - remove delay as option and hard-code it, or:
2. Leave the option, but ensure that is is at least 100, if it is lower - set it to 100 and warn user in console

daun commented:
If it depends on how busy the screen reader is, it probably has to do with the live region level. It's `polite` by default, meaning it won't interrupt any ongoing announcements. Switching the `level` to `assertive` should solve the issue without the need for a delay.

```js
const liveRegion = new OnDemandLiveRegion({ level: 'assertive' })
```

Whether this plugin should be `assertive` by default is probably a matter of debate. We might just make the live region library itself configurable by adding a `liveRegion` option that just passes everything on to the live region instance. That way, people are free to tweak these according to their requirements.

ScoobyDid commented:
@daun Good point; it looks like a good solution, but i have tested it on NVDA and this is what i got:

- With `assertive`  level plugin also adds `role="alert"`, which makes screen reader say "Alert Navigated to..." instead of "Navigated to"
- Plugin does not allow to control role as option, but if i edit plugin files and remove `role="alert"`, or set  `role="status"`, while keeping `aria-live="assertive"`, it works exactly the same as with `aria-live="polite"` - often not being pronounced

Overall, if it is ok to have "Alert Navigated to..." instead of "Navigated to", it works well.

I tested on: NVDA, windows 11, chrome 121, 'guest' window

Also, while testing, i encountered another problem - previously i have been testing with `contentSelector` option set to `body` (as browsers would handle focus on traditional navigation), and it worked well, but when i changed `contentSelector` to `main` (or any other element excluding body), there were 2 issues:

1. Updated text often had been pronounced after content of focused element
2. Again, sometimes content has not been pronounced at all

I fixed first problem by adding setTimeout before setting focus to element, that is >= delay, and second problem by increasing delay to 200. I am not sure if it is the most reliable solution, but i didn't encounter any problems with it so far.

daun commented:
This is super helpful. When creating this plugin, I only had VoiceOver on Mac/iOS to test the announcements. We should definitely make sure that the contents are announced correctly.

Having "Alert:" prefixed to the announcement seems a bit harsh, let's try without that 🤠

I'd like to do some more research and test your suggested fix on some other browser/reader combinations to get a grasp of the issue. Ideally, we can fix this without adding timeouts throughout the code, but if we need to, we will.

daun commented:
@ScoobyDid Just a heads-up that your contribution hasn't fallen through the cracks. I'm still figuring out how to properly test this on a Windows device, so please bear with us here.

ScoobyDid commented:
@daun Excuse me for the late reply. It's totally ok, it's not urgent for me right now. Take your time, and please let me know if there's anything I can assist with in the meantime. I think it would be convenient to use browserstack`s [accessibility testing feature](https://www.browserstack.com/accessibility-testing?utm_source=Email&utm_medium=Email&utm_campaign=FreeTrial_Verification) - you can sign up for a free trial (if you don't have an account), and use different screen readers there, including NVDA

daun commented:
@ScoobyDid Thanks for the hint about Browserstack, I do have an account with them but I didn't know about the screen reader feature. Turns out that feature costs 249$ a month, haha. They seem to have an open source plan that's free, I'll go do some research on that.

ScoobyDid commented:
Wow 249$ is... unexpected 😂. As i know, they have a free 15-day free trial separately for accessibility testing, but if they have free open source plan, that's incredible.

ScoobyDid commented:
@daun That is fantastic! I'm glad my input can be useful. Absolutely, feel free to add me to the list of contributors. Looking forward to seeing how the updates unfold, even if it takes a few separate PRs. Keep up the great work! 😊

hirasso commented:
There is also some great news from the test tools front: swup was granted with a free-for-open-source account at [Assistiv Labs](https://assistivlabs.com/)! That means we now have access to testing swup with [NVDA](https://github.com/nvaccess/nvda) 🤠

[JAWS](https://www.freedomscientific.com/products/software/jaws/) is not included due to their [ridiculous licensing costs and business practices](https://netz-barrierefrei.de/en/jaws-nvda.html). 

But from what I was able to find online, it looks like it should work in JAWS if it's working in NVDA, fingers crossed.

daun commented:
@ScoobyDid Any particular way you'd like to be mentioned in the contributor list? GitHub username and link?

ScoobyDid commented:
@daun Kindly include my current GitHub username along with a link to my profile (https://github.com/ScoobyDid). No specific preferences beyond that, thank you! ✨
